File: src/fddb.cpp
#include <cstdio>
#include <iostream>
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include "jfda.hpp"

using namespace cv;
using namespace std;
using namespace jfda;

int main(int argc, char *argv[]) {
  const string fddb_dir = "../data/fddb";
  const string prefix = fddb_dir + "/images/";
  const string result_prefix = fddb_dir + "/result/out";

  jfda::Detector detector;

  for (int i = 1; i <= 10; i++) {
    char fddb[300];
    char fddb_out[300];
    char fddb_answer[300];

    printf("Testing FDDB-fold-%02d.txt\n", i);
    sprintf(fddb, "%s/FDDB-folds/FDDB-fold-%02d.txt", fddb_dir.c_str(), i);
    sprintf(fddb_out, "%s/result/fold-%02d-out.txt", fddb_dir.c_str(), i);
    sprintf(fddb_answer, "%s/FDDB-folds/FDDB-fold-%02d-ellipseList.txt", fddb_dir.c_str(), i);

    FILE* fin = fopen(fddb, "r");
    FILE* fanswer = fopen(fddb_answer, "r");
    if (!fin || !fanswer) {
      printf("Can not open file\n");
      exit(-1);
    }

#ifdef WIN32
    FILE* fout = fopen(fddb_out, "wb"); // replace \r\n on Windows platform
#else
    FILE* fout = fopen(fddb_out, "w");
#endif // WIN32

    char buff[300];
    char _buff[30];
    char path[300];

    int counter = 0;
    while (fscanf(fin, "%s", path) > 0) {
      string full_path = prefix + string(path) + string(".jpg");
      Mat img = imread(full_path, CV_LOAD_IMAGE_COLOR);
      if (!img.data) {
        printf("Can not open %s, Skip it", full_path.c_str());
        continue;
      }

      double fps = 0.;
      vector<jfda::FaceBBox> result = detector.detect(img, 3);
      const int n = result.size();

      fprintf(fout, "%s\n%d\n", path, n);

      for (int j = 0; j < n; j++) {
        const FaceBBox& r = result[j];
        fprintf(fout, "%d %d %d %d %lf\n", r.x, r.y, r.w, r.h, r.score);
      }

      counter++;
      sprintf(buff, "%s/%s.jpg", result_prefix.c_str(), path);

      // get answer
      int face_n = 0;
      fscanf(fanswer, "%s", path);
      fscanf(fanswer, "%d", &face_n);
      for (int k = 0; k < face_n; k++) {
        double major_axis_radius, minor_axis_radius, angle, center_x, center_y, score;
        fscanf(fanswer, "%lf %lf %lf %lf %lf %lf", &major_axis_radius, &minor_axis_radius, \
                                                    &angle, &center_x, &center_y, &score);
        // draw answer
        angle = angle / 3.1415926*180.;
        cv::ellipse(img, Point2d(center_x, center_y), Size(major_axis_radius, minor_axis_radius), \
                    angle, 0., 360., Scalar(255, 0, 0), 2);
      }

      // draw result
      for (int j = 0; j < n; j++) {
        const FaceBBox& r = result[j];
        double score = r.score;
        cv::rectangle(img, Rect(r.x, r.y, r.w, r.h), Scalar(0, 0, 255), 3);
        // draw score
        sprintf(_buff, "%.4lf", score);
        cv::putText(img, _buff, cv::Point(r.x, r.y), cv::FONT_HERSHEY_PLAIN, 1, cv::Scalar(0, 255, 0), 2);
        //// draw shape
        //for (int k = 0; k < 5; k++) {
        //  cv::circle(img, Point(r.landmark[2 * k], r.landmark[2 * k + 1]), 3, Scalar(0, 255, 0), -1);
        //}
      }

      string fname(path);
      std::replace(fname.begin(), fname.end(), '/', '_');
      fname = result_prefix + "/" + fname + ".jpg";
      cv::imwrite(fname, img);
    }
  }

  return 0;
}

File: fddb.py
#!/usr/bin/env python2.7
# pylint: disable=bad-indentation, no-member, invalid-name, line-too-long

import argparse
import cv2
import caffe
from jfda.config import cfg
from jfda.utils import get_logger, Timer
from jfda.detector import JfdaDetector


def main(args):
  if args.net == 'p':
    net = ['proto/p.prototxt', 'model/p.caffemodel']
  elif args.net == 'r':
    net = ['proto/p.prototxt', 'model/p.caffemodel',
           'proto/r.prototxt', 'model/r.caffemodel']
  else:
    net = ['proto/p.prototxt', 'model/p.caffemodel',
           'proto/r.prototxt', 'model/r.caffemodel',
           'proto/o.prototxt', 'model/o.caffemodel',]
  detector = JfdaDetector(net)
  if args.pnet_single:
    detector.set_pnet_single_forward(True)
  logger = get_logger()
  counter = 0
  timer = Timer()
  param = {
    'ths': [0.5, 0.5, 0.4],
    'factor': 0.709,
    'min_size': 24,
  }
  total_time = 0.
  total_width = 0
  total_height = 0
  for i in range(10):
    logger.info('Process FOLD-%02d', i)
    txt_in = cfg.FDDB_DIR + '/FDDB-folds/FDDB-fold-%02d.txt'%(i + 1)
    txt_out = cfg.FDDB_DIR + '/result/fold-%02d-out.txt'%(i + 1)
    answer_in = cfg.FDDB_DIR + '/FDDB-folds/FDDB-fold-%02d-ellipseList.txt'%(i + 1)
    fin = open(txt_in, 'r')
    fout = open(txt_out, 'w')
    ain = open(answer_in, 'r')
    for line in fin.readlines():
      line = line.strip()
      in_file = cfg.FDDB_DIR + '/images/' + line + '.jpg'
      out_file = cfg.FDDB_DIR + '/result/images/' + line.replace('/', '-') + '.jpg'
      counter += 1
      img = cv2.imread(in_file, cv2.IMREAD_COLOR)
      timer.tic()
      bboxes = detector.detect(img, **param)
      timer.toc()
      total_time += timer.elapsed()
      h, w = img.shape[:-1]
      total_height += h
      total_width += w
      logger.info('Detect %04d th image costs %.04lfs, average %.04lfs, %d x %d', counter,
                  timer.elapsed(), total_time / counter, total_height / counter, total_width / counter)
      fout.write('%s\n'%line)
      fout.write('%d\n'%len(bboxes))
      for bbox in bboxes:
        x1, y1, x2, y2, score = bbox[:5]
        fout.write('%lf %lf %lf %lf %lf\n'%(x1, y1, x2 - x1, y2 - y1, score))
      # draw ground truth
      ain.readline() # remove fname
      n = int(ain.readline().strip())
      for i in range(n):
        line = ain.readline().strip()
        components = [float(_)  for _ in line.split(' ')[:5]]
        major_axis_radius, minor_axis_radius, angle, center_x, center_y = components
        angle = angle / 3.1415926 * 180.
        center_x, center_y = int(center_x), int(center_y)
        major_axis_radius, minor_axis_radius = int(major_axis_radius), int(minor_axis_radius)
        cv2.ellipse(img, (center_x, center_y), (major_axis_radius, minor_axis_radius), angle, 0, 360, (255, 0, 0), 2)
      # draw and save
      for bbox in bboxes:
        x1, y1, x2, y2, score = bbox[:5]
        x1, y1, x2, y2 = int(x1), int(y1), int(x2), int(y2)
        cv2.rectangle(img, (x1, y1), (x2, y2), (0, 0, 255), 2)
        cv2.putText(img, '%.03f'%score, (x1, y1), cv2.FONT_HERSHEY_PLAIN, 1, (0, 255, 0))
        landmark = bbox[9:].reshape((5, 2))
        for x, y in landmark:
          x, y = int(x), int(y)
          cv2.circle(img, (x, y), 2, (0, 255, 0), -1)
      cv2.imwrite(out_file, img)
    fin.close()
    fout.close()
    ain.close()


if __name__ == '__main__':
  parser = argparse.ArgumentParser()
  parser.add_argument('--gpu', type=int, default=-1, help='gpu id to use, -1 for cpu')
  parser.add_argument('--net', type=str, default='o', help='net type to do fddb benchmark')
  parser.add_argument('--pnet-single', action='store_true', help='pnet use single forward')
  args = parser.parse_args()
  assert args.net in ['p', 'r', 'o']

  print args

  if args.gpu >= 0:
    caffe.set_mode_gpu()
    caffe.set_device(args.gpu)
  else:
    caffe.set_mode_cpu()

  main(args)
